### PR TITLE
refactor!: Moves the email IDP into the IDP pattern and adds tests for everything.

### PR DESCRIPTION
**Pull request by SandPod**
This is a major PR that includes multiple things:

- Implements the IDP structure for the email IDP provider as mentioned in https://github.com/serverpod/serverpod/pull/4064
- Adds tests for everything exposed by the email IDP provider.
- Fixes several bugs and issues found when implementing tests.
- Does an initial ux pass on the methods exposed by the provider.
- Adds the provider to the auth example project.
- Modifies a database model related to completion attempts.

This PR can be reviewed commit by commit but best is probably to simply observe the end state the project ends up in.

I would suggest that minor comments are fixed in follow up PRs if the overall structure of the interfaces looks good.

Fixes: #4052 

## Pre-launch Checklist

- [ ] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [ ] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [ ] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [ ] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes
Yes, refactor the email auth provider for the new auth. Contains multiple breaking changes.

### CODE REVIEW

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/providers/apple.dart [hunk 1]
```diff
@@ -2,4 +2,5 @@
 /// Serverpod IDP module.
 library;
 
+export '../src/common/auth_services.dart';
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/providers/apple.dart:5]: Question: This export here is intentional? Why only on Apple and Email, but not on Google? Or should it be only on the most external barrel file?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp_admin.dart [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    );
+  }
+
+  /// Checks whether an email authentication exists for the given email address.
+  Future<
+      ({
+        UuidValue authUserId,
+        UuidValue emailAccountId,
+        bool hasPassword,
+      })?> findAccount(
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp_admin.dart]: Question: Since all other returns were replaced by actual classes, has this record been left here intentionally?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        );
+
+        passwordResetRequestId = await session.db.transaction(
+          (final transaction) => fixture.emailIDP.startPasswordReset(
+            session,
+            email: email,
+            transaction: transaction,
+          ),
+        );
+      });
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Question: Since only one operation is done in the transaction, could this be a direct call to `startPasswordReset`?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+    });
+
+    tearDown(() async {
+      await fixture.tearDown(session);
+    });
+
+    test(
+        'when finishPasswordReset is called with valid parameters then it throws EmailAccountPasswordResetException with expired reason',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Suggestion: Add another test with invalid parameters to assert that the endpoint throws `invalid` instead of `expired` to avoid leaking if the request exists.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+          ),
+        ),
+      );
+    });
+  });
+
+  // TODO: This scenario is a bit strange since it allows the password reset
+  // to be completed even though the auth user is blocked and just prevents
+  // a new session from being created.
+  withServerpod('Given password reset request exists for blocked auth user',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Question: Should we spin up an issue instead of leaving this only as `TODO` here?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        );
+      });
+    });
+
+    tearDown(() async {
+      await fixture.tearDown(session);
+    });
+
+    test(
+        'when finishRegistration is called with valid parameters then it throws EmailAccountRequestException with expired reason',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart]: Suggestion: Also missing a test here regarding the use of invalid verification code on expired request to ensure we throw `invalid` instead of `expired`.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        );
+
+        await expectLater(
+          result,
+          throwsA(isA<AuthUserBlockedException>()),
+        );
+      });
+    },
+  );
+  withServerpod(
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart:124]: Suggestion(NIT): Missing newline between test groups.

```suggestion
  );
  
  withServerpod(
```

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+      final sessions = await AuthSessions.admin.findSessions(
+        session,
+        authUserId: authUserId,
+      );
+
+      expect(sessions, hasLength(1));
+      expect(sessions.single.authUserId, authSuccess.authUserId);
+    });
+  });
+}
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart:537]: Discussion: Finish password reset tests are not covering `tooMany` scenarios, but we are still to discuss whether we should separate this step into verification first, set password last. Should we add the `tooMany` here either way or avoid the rework and add such tests only after(/if) we split the method?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+          isA<EmailAccountRequestException>().having(
+            (final e) => e.reason,
+            'reason',
+            EmailAccountRequestExceptionReason.invalid,
+          ),
+        ),
+      );
+    });
+  });
+}
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart:237]: Finish account registration tests are missing coverage on `tooMany` exceptions to ensure that such exceptions are being correctly replaced by `invalid` to hide if the request exists. The test must cover at least two scenarios: (1) with a valid and (2) an invalid request ID. The edge case is when the user completes with the correct code for an existing request after trying too many times (or when the right one is the exceeding one).

Comment (marcelomendoncasoares) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart:237]: Discussion: In the edge case above (right code after too many attempts), the ideal would be to actually inform `tooMany` to the user. This would be safe because the right verification code was used, such as how we do with `expired`. Saying `tooMany` would certainly be more informative to the UI, so we can tell the user to generate a new request instead of leaving them in doubt regarding where the error is. @SandPod, do you think we could make this change?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+            newPassword: allowedNewPassword,
+            transaction: transaction,
+          ),
+        );
+
+        await expectLater(result, completion(isA<AuthSuccess>()));
+      });
+
+      test(
+          'when finishPasswordReset is called with invalid verification code then it throws EmailAccountPasswordResetException',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]:
```suggestion
          'when finishPasswordReset is called with invalid verification code then it throws EmailAccountPasswordResetException with invalid reason',
```

Comment (marcelomendoncasoares) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Suggestion(NIT): The suggestion above follows the `with <reasonName> reason` pattern and the same used on the finish registration, but reads confusing - since it is not _an invalid reason_, but _the reason "invalid"_. We could change all to use quotes on the name of the reasons or maybe change the order on these ones - `with reason invalid` - to remove ambiguity.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+          // Expected
+        }
+      });
+
+      tearDown(() async {
+        await fixture.tearDown(session);
+      });
+
+      test(
+          'when login is called with valid credentials then it throws EmailAccountLoginException with tooManyAttempts',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart]: Missing a test case here to ensure symmetry on the response when using invalid password, which must also throw `tooManyAttempts`.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+        await expectLater(
+          result,
+          throwsA(isA<AuthUserBlockedException>()),
+        );
+      });
+    },
+  );
+  withServerpod(
+    'Given email account with invalid logins matching rate limit',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart:125]: Also missing another scenario with invalid email account to ensure rate limiting is being applied for inexistent accounts - and not leaking if the account exists.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        session = sessionBuilder.build();
+        fixture = EmailIDPTestFixture();
+      });
+
+      tearDown(() async {
+        await fixture.tearDown(session);
+      });
+
+      test(
+          'when startPasswordReset is called then it returns dummy uuid without throwing',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart]: We should also add a test pinning down the version of the UUID to ensure that we respond with the correct type of random UUID. Otherwise, it could leak existing accounts due to a difference between the model and the random generation on the code.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            email: 'nonexistent@serverpod.dev',
+            transaction: transaction,
+          ),
+        );
+
+        await expectLater(result, completion(isA<UuidValue>()));
+      });
+    },
+  );
+}
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart:248]: We are missing the `tooMany` tests on the start password reset to ensure that rate limiting is being applied without revealing that accounts exist or not.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+          password: EmailAccountPassword.fromString(password),
+        );
+      });
+
+      tearDown(() async {
+        await fixture.tearDown(session);
+      });
+
+      test(
+          'when start registration is called for the same user then it returns dummy password reset request id',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]: If we decide to move the UUID version comparison tests as unit tests, we should also add one here.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+            (final transaction) => fixture.emailIDP.startRegistration(
+              session,
+              email: email,
+              password: password,
+              transaction: transaction,
+            ),
+          );
+        });
+
+        test('then it returns password reset request id', () async {
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]:
```suggestion
        test('then it returns account registration request id', () async {
```

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+          await expectLater(
+              accountRequestIdFuture, completion(isA<UuidValue>()));
+        });
+
+        test('then password reset request can be used to complete registration',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]:
```suggestion
        test('then account registration request can be used to complete registration',
```

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+              transaction: transaction,
+            ),
+          );
+
+          await expectLater(authSuccessFuture, completion(isA<AuthSuccess>()));
+        });
+      });
+    },
+  );
+}
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:169]: We are also missing the `tooMany` tests on the account request to ensure that rate limiting is being applied without revealing that accounts exist or not.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp_admin_test.dart [hunk 19]
```diff
@@ -0,0 +1,573 @@
...
+            transaction: transaction,
+          ),
+        );
+
+        expect(result, isNotNull);
+        expect(result?.hasPassword, isTrue);
+      });
+
+      test(
+          'when findAccount is called with uppercase email then it finds email account',
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp_admin_test.dart]: Should this test be moved to a different group? Since it has no relation with the user having a password set.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart [hunk 20]
```diff
@@ -0,0 +1,406 @@
...
+      // Verify the session was destroyed by checking if it still exists
+      final sessions = await AuthSessions.admin.findSessions(
+        session,
+        authUserId: authUserId,
+      );
+
+      expect(sessions, hasLength(1));
+      expect(sessions.single.authUserId, authSuccess.authUserId);
+    });
+  });
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart:392]: There are two test cases that are not covered here:
- `Given finished password reset attempt when trying to authenticate with old credentials then it throws EmailAccountLoginException with invalidCredentials`
- `Given deleted account with a pending reset request when trying to complete the request then it throws EmailAccountPasswordResetException with invalid reason`

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+              verificationCode: verificationCode,
+              transaction: transaction,
+            ),
+          );
+
+          await expectLater(authSuccessFuture, completion(isA<AuthSuccess>()));
+        });
+      });
+    },
+  );
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:168]: One test case that is not covered here:

- (Given no email account) `when startRegistration is called with invalid email address then it throws (?)` - Not sure what is the current behavior, but we should be informative that the email is not a valid address. It does not leak information and allow a good UI feedback to users.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+            session,
+            email: email,
+            password: password,
+            transaction: transaction,
+          ),
+        );
+
+        await expectLater(result, completion(isA<UuidValue>()));
+      });
+    },
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:109]: One test case that is not covered here:

- (Given an existing email account) `when startRegistration is called with the email address in uppercase then it returns dummy account registration request id`

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 15 above)
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]:
```suggestion
          'when start registration is called for the same user then it returns dummy account registration request id',
```

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart (hunk 13 above)
Comment (marcelomendoncasoares) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart]: This test exists on the integration test on the test project, but should it be here instead? Feels more like a unit test that should be asserted close to the package rather than on the `serverpod_new_auth_test_server` test project. The file that contains this test today is: `serverpod_new_auth_test_server/test/integration/email_account_endpoint_test.dart`.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp.dart [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+          // one used by the [EmailAccountRequest] model to prevent attackers from
+          // using the difference on the version bit of the uuid to determine whether
+          // an email is registered or not.
+          return result.accountRequestId ?? const Uuid().v4obj();
+        },
+      ),
+    );
+  }
+
+  Future<AuthSuccess> _createSession(
```
Comment (marcelomendoncasoares) [inline modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp.dart:221]: Question: Should we rename this method to something more generic as a preparation to receive the new token issuers architecture? Or do we expect this method to be entirely removed by then, so it has not been touched?

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/providers/apple.dart (hunk 1 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/providers/apple.dart:5]: All three of them contain this export at the moment. It will be removed with the new AuthConfig.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp_admin.dart (hunk 2 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp_admin.dart]: Now returns the account: [0a182c4](https://github.com/serverpod/serverpod/pull/4089/commits/0a182c47d9f066e3798816f80367b1ab55da7486)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 3 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Good catch, yes it could be! 👍 Will fix this here and for other places.

Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Fixed: [9daea7e](https://github.com/serverpod/serverpod/pull/4089/commits/9daea7ec20a8fecb74d8cad376cb993eb32f41a7)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 4 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Added here: [7310905](https://github.com/serverpod/serverpod/pull/4089/commits/731090574647656ed367cbc18a0029751cf70b28)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 5 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: You are right! @coderabbitai - Could you create an issue for this?

Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Removed the comment: [e30dfaa](https://github.com/serverpod/serverpod/pull/4089/commits/e30dfaad7ac31e338dd6751a78d73a737778e3df)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart (hunk 6 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart]: Fixed: [1175721](https://github.com/serverpod/serverpod/pull/4089/commits/11757216d2172a16e33d1a0b799e51270fc4c68e)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart (hunk 7 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart:124]: Fixed: [b9c19bb](https://github.com/serverpod/serverpod/pull/4089/commits/b9c19bbb27caa03fc5c93b2d44882d1e3339d8e4)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 8 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart:537]: I think it is good to have this validation either way so I added the test: [6080813](https://github.com/serverpod/serverpod/pull/4089/commits/6080813f7eefd12c1a1615167fcfad562b77d86b)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart (hunk 9 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart:237]: Since we have expired, I think this is a reasonable change to make (and apply to finish password reset as well), but the question then becomes whether we should have expired.

Right now, our security is that a matching UUID for an account request is almost impossible to get right, given the nature of UUIDs. So we assume that as long as you have the request ID, you are, in fact, aware that a request is underway. Meaning that we should safely be able to provide the client with client-enhancing details.

So my suggestion is that  we expose `tooMany` for both `finish_registration` and `finish_password_reset`.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 10 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart]: Fixed here, moved the phrasing an placed the reason in `""` for all cases: [2a3d4c0](https://github.com/serverpod/serverpod/pull/4089/commits/2a3d4c0e31a8b5592cec6c3712f74cd8e0025078)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart (hunk 11 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart]: Fixed: [07d1ba5](https://github.com/serverpod/serverpod/pull/4089/commits/07d1ba50991373c36fa8f6547a0a596f9d8e3aa8)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart (hunk 12 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_login_test.dart:125]: Fixed: [e6f6b46](https://github.com/serverpod/serverpod/pull/4089/commits/e6f6b4661111c6700dd2f10ebf254e17c1d1e466)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart (hunk 13 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart]: Moved the tests into unit tests: [119393a](https://github.com/serverpod/serverpod/pull/4089/commits/119393a20fc8052901245fbb3b55fe6b0744e1f6)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart (hunk 14 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_password_reset_test.dart:248]: fixed: [1ae4467](https://github.com/serverpod/serverpod/pull/4089/commits/1ae4467b6e6ea5f51571d66d682586f413d59668) (These might be patched later if we change the behavior.)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 15 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]: Fixed with: [119393a](https://github.com/serverpod/serverpod/pull/4089/commits/119393a20fc8052901245fbb3b55fe6b0744e1f6)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 16 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]: Fixed: [c9f1053](https://github.com/serverpod/serverpod/pull/4089/commits/c9f1053ae11a091264450af8962f429691c41fa9)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 17 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]: Fixed: [c9f1053](https://github.com/serverpod/serverpod/pull/4089/commits/c9f1053ae11a091264450af8962f429691c41fa9)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 18 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:169]: There is no rate limit in this area.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp_admin_test.dart (hunk 19 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp_admin_test.dart]: Fixed: [8bad007](https://github.com/serverpod/serverpod/pull/4089/commits/8bad007de89a7c5ff51ab011aea2f46493d9807e)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart (hunk 20 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart:392]: We don't actually support deleting accounts at the moment, that should probably be added, but we can do that in a separate PR. @coderabbitai can you create an issue for adding account deletion to email_idp_admin.dart?

Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_password_reset_test.dart:392]: Test added for first case here: [f3113c5](https://github.com/serverpod/serverpod/pull/4089/commits/f3113c5101f07d9c028129c04b73fcf7cc64f53c)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 21 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:168]: This currently does not throw anything but just returns a dummy uuid. Let's discuss how we should modify this piece of code. I believe exceptions is probably the right way, return values seems a bit strange for the case here.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 22 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:109]: Fixed: [cda2ff1](https://github.com/serverpod/serverpod/pull/4089/commits/cda2ff181bbd07c813ac6298eaadb9e6014af0b0)

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 15 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart]: Fixed.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp.dart (hunk 23 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/lib/src/providers/email/business/email_idp.dart:221]: It will be entirely removed so I didn't bother renaming it at this stage.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart (hunk 9 above)
Comment (marcelomendoncasoares) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_finish_registration_test.dart:237]: @coderabbitai Please spin up an issue regarding this change, so we don't bloat this PR with business logic changes.

**Comment by SandPod** [comment]
@coderabbitai When reviewing this PR, we noticed that we are writing attempts before checking if the user has passed their allowed threshold. This will cause issues because users attempting again will prolong their blocked period, and we will add unnecessary writes to the database.

Please create an issue for fixing this.

File: modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart (hunk 21 above)
Comment (SandPod) [reply modules/new_serverpod_auth/serverpod_auth_idp/serverpod_auth_idp_server/test/email/email_idp/email_idp_start_registration_test.dart:168]: @coderabbitai create an issue for modifying the return type to throw if the email is invalid or if password violates password policy.

**Review by marcelomendoncasoares — APPROVED** [review]
LGTM!
